### DISCUSSION TITLE
Dynamic inclusion of global resources in _host.cshtml

### DISCUSSION THREAD

**Discussion by ADefWebserver** [oqtane/oqtane.framework#1011] [Q&A] [answered]
- (Primarily asking for guidance from @sbwalker but anyone who has code examples/experience to offer is welcome :))
- (This is related to: [In Progress: 3rd Party Components Integration in Blazor Oqtane using IServerStartup and IClientStartup](https://github.com/oqtane/oqtane.framework/discussions/1008))

It appears from [this thread](https://github.com/oqtane/oqtane.framework/discussions/690#discussioncomment-58579) that the proper way to include references to JavaScript in the Oqtane _Hosts file is to use the [IHostResources interface](https://github.com/oqtane/oqtane.framework/pull/714).

It appears that this will be required to implement support in Oqtane for Blazor library components from [Radzen ](https://razor.radzen.com/) and [Syncfusion](https://www.syncfusion.com/blazor-components).

Before I get started on this, I was hoping to find **any** example code. Thanks!

**Comment by larsontim12** [comment]
I have a support ticket into Syncfusion on integrating their controls into an external module for Oqtane without requiring any changes to the Oqtane framework.  They have been working on it and having difficulties getting the sample to render it comes up with a run-time error (they haven't sent me the sample yet.)  I will let you know when they get back to me with a sample.

I know you have worked with Syncfusion and might be able to get in touch with them to see where they stand on it.

I think a great example would be to port your silverlight expense report DNN module to Oqtane.  I appreciate you samples and tutorials!

Thnaks -Tim

**Comment by sbwalker** [comment, marked as answer]
@ADefWebserver there are a couple of options for registering global resources:

You can declare the resources in each razor component for your module and set the Declaration property to Global:

```
    public override List<Resource> Resources => new List<Resource>()
    {
        new Resource { ResourceType = ResourceType.Stylesheet, Declaration = ResourceDeclaration.Global, Url = ModulePath() + "file.css" },
        new Resource { ResourceType = ResourceType.Script, Declaration = ResourceDeclaration.Global, Url = ModulePath() + "file.js" }
    };
```

Or you can create a class in your external module's Server project:

```
 public class HostResources : IHostResources
    {
        public List<Resource> Resources => new List<Resource>()
        {
            new Resource { ResourceType = ResourceType.Stylesheet, Url = "Modules/" + GetType().Namespace + "/file.css" },
            new Resource { ResourceType = ResourceType.Script, Url = "Modules/" + GetType().Namespace + "/file.js" }
        };
    }
```

The benefit of IHostResources is that it can be declared in one place in your module rather than in each razor component - but both approaches accomplish the same thing.

**Reply by ADefWebserver** [reply]
@sbwalker - Thanks. This is what I was looking for. I think I will do a separate Blog post on this and then link back to it for the planned Blog posts on integrating the 3rd party libraries using the proof of concept you cover here: https://github.com/oqtane/oqtane.framework/discussions/690#discussioncomment-58150

**Reply by ADefWebserver** [reply]
@sbwalker I made a blog post covering IHostResources here: https://blazorhelpwebsite.com/ViewBlogPost/45 I will do another Blog post on integrating the 3rd party libraries next week.

**Comment by ADefWebserver** [comment]
@larsontim12 - Thank you for letting me know that you have been in contact with Syncfusion. However, I won’t be surprised if they end up contacting **me** for assistance on this issue :) - - I think we have the --pieces-- needed to make all this work. I just need final guidance from @sbwalker before I "dive" in over the next week.

I will probably end up doing a series of articles...

**Reply by ADefWebserver** [reply]
@larsontim12 - I tried to make Syncfusion work, but, I could not :(

My sample code is here: [BlazorHelpWebsite.SyncfusionTest.zip](https://github.com/oqtane/oqtane.framework/files/5747818/BlazorHelpWebsite.SyncfusionTest.zip)

Here is what I tried:

![image](https://user-images.githubusercontent.com/1857799/103224843-671f2a00-48dd-11eb-968c-236c3b6a2b6a.png)
![image](https://user-images.githubusercontent.com/1857799/103224865-743c1900-48dd-11eb-8471-f78b04ec19db.png)
![image](https://user-images.githubusercontent.com/1857799/103224888-86b65280-48dd-11eb-99a1-a550ce076b69.png)
![image](https://user-images.githubusercontent.com/1857799/103224901-90d85100-48dd-11eb-8925-fa4b62de5c98.png)

**Reply by larsontim12** [reply]
@ADefWebserver  I heard back from Syncfusion just recently.  This is what they said:

We have validated the oqtane with syncfusion external modules. We can able to render our Syncfusion components as external modules in oqtane. But here we will face issues with referencing to the scripts and styles since our syncfusion controls use the staticWebAssets from the package. However, we can overcome this by using CRG/CDN generated scripts and styles references. This is applicable for 18.2.x - 18.3.38 version syncfusion packages only.

From, 18.3.40 version packages we disabled the local scripts referencing since the scripts are loaded internally using JavaScript Isolation. Please refer to the below release notes for further details.

https://blazor.syncfusion.com/documentation/release-notes/18.3.40/?type=all#common

So, the local and CDN based scripts referencing is not recommended from 18.3.40.

For now, we need to load these static assets from external modules to oqtane framework. This would be done by oqtane end.

We hope this would be addressed on their side as per this discussion and you also part of it.

https://github.com/oqtane/oqtane.framework/discussions/693

https://github.com/oqtane/oqtane.framework/discussions/1011

However, we have attached the sample in which we are referencing our CDN styles and CRG generated scripts(using v18.2.59 package). 

But this would be not worked with our latest 18.4.x version packages as we stated above.

**Reply by ADefWebserver** [reply]
@larsontim12  - Thanks for this! - But this seems to be a "showstopper" unless we can get Oqtane to: "load static assets from external modules". I have no idea why that currently doesn't work, or how to do that :(  :( (cc: @sbwalker) 

This thread discusses this issue: https://github.com/oqtane/oqtane.framework/discussions/764

**Reply by sbwalker** [reply]
@ADefWebserver @larsontim12 I am confused why Syncfusion believes this is an Oqtane issue. Oqtane simply renders standard Blazor components and Blazor components have the full freedom on how they wish to interact with JavaScript libraries. I have looked at the JavaScript Isolation feature in .NET 5 and it does have different characteristics than .NET Core 3.1: 

https://docs.microsoft.com/en-us/aspnet/core/blazor/call-javascript-from-dotnet?view=aspnetcore-5.0#blazor-javascript-isolation-and-object-references

First, it requires that you use JavaScript Modules ( which is not supported by many JavaScript libraries currently ). Based on Syncfusion's response it sounds like they do support JavaScript Modules. Second, you must import the JavaScript Modules and declare its location in your code:

```
var module = await js.InvokeAsync<IJSObjectReference>(
    "import", "./_content/MyComponents/exampleJsInterop.js");
```

In the example above it is expecting it to be located in _content but there is nothing stopping you from specifying any valid location. So I am trying to understand what Syncfusion means when they say "scripts are loaded internally using JavaScript Isolation"? Perhaps it means that within the Syncfusion components there are hardcoding a path where they expect the JavaScript Modules to exist?

**Reply by ADefWebserver** [reply]
@sbwalker - This week, let me try to update my Article [Using Custom JavaScript in Blazor Oqtane](https://blazorhelpwebsite.com/ViewBlogPost/45) to show an example of pulling the JavaScript from "_content" (rather than requiring it to be registered in _host.cshtml).

Then it would be clear to Syncfusion (and others).

**Reply by sbwalker** [reply]
@ADefWebserver I think the real issue may be included in this description of static assets in Razor Class Libraries ( Oqtane modules are essentially RCLs ):

https://docs.microsoft.com/en-us/aspnet/core/razor-pages/ui-class?view=aspnetcore-5.0&tabs=visual-studio#create-an-rcl-with-static-assets

During development Microsoft recommends putting static assets in the wwwroot folder ( which Oqtane follows ). However when an RCL is packed into a Nuget package using dotnet pack, it’s assets are extracted to the  _content folder ( whereas Oqtane extracts to wwwroot ). I am guessing that Syncfusion’s JavaScript files end up in _content... and it’s components include a hard coded reference to this location - which causes the issue. There is also the matter of “.” support in folder paths which I need to research further.

**Reply by sbwalker** [reply]
https://github.com/dotnet/aspnetcore/issues/6349#issuecomment-497672169

javiercn:
“... you just need to put the files from the library inside a wwwroot folder and they will get exposed under the _content/<<library>>/ path.
For example, if your library name is MyLibrary and you have in your library wwwroot\sample.js it will get exposed on consuming apps at _content/mylibrary/sample.js (the contents of your library wwwroot folder will end up when you publish your app in the wwwroot folder of the published app, under _content/mylibrary”

**Reply by ADefWebserver** [reply]
@sbwalker - I am not able to get **Oqtane** to pull any **.css** or **.js** resource that expects to be at "**_content/**".
For example, with **Radzen**, If I create a non-**Oqtane** project following these directions: https://razor.radzen.com/get-started

The links point to "**_content/**" and it works:

![image](https://user-images.githubusercontent.com/1857799/103484081-faa2a000-4da0-11eb-9bcf-fbda9ebd1da5.png)
![image](https://user-images.githubusercontent.com/1857799/103484087-02624480-4da1-11eb-937c-7cc7325e4446.png)

But, with **Oqtane**, I can register the links correctly:

![image](https://user-images.githubusercontent.com/1857799/103484112-36d60080-4da1-11eb-98a2-e37d8b44d892.png)
![image](https://user-images.githubusercontent.com/1857799/103484121-45241c80-4da1-11eb-9ff7-0cc24d8f1424.png)

But they won't load:

![image](https://user-images.githubusercontent.com/1857799/103484134-51a87500-4da1-11eb-8a51-b99e73c0748f.png)

With **Radzen**, we can get around it by [extracting the resources and placing them in the module](https://blazorhelpwebsite.com/ViewBlogPost/46), but, **Syncfusion** (no longer) allows you to override the location of the resources, _they have to be at_ "_content/".

So it appears to be a showstopper to me :(

**Reply by sbwalker** [reply]
@ADefWebserver I can see that you registered the JavaScript in your Oqtane module... and you specified the relative Url to the _content folder... but how did you deploy the JavaScript files from your module to the Oqtane runtime environment? Based on the fact the requests for these JavaScript files are throwing 404 errors means that those JavaScript files do not exist at the Url specified. And I am not sure how they would exist at that location unless you took some action to deploy them there. So how did you configure your Oqtane module to deploy the static assets for Radzen... did you use the debug.cmd in the Deploy project in your external module... or did you create a Nuget package and deploy it?

**Reply by ADefWebserver** [reply]
> Based on the fact the requests for these JavaScript files are throwing 404 errors means that those JavaScript files do not exist at the Url specified. And I am not sure how they would exist at that location unless you took some action to deploy them there. So how did you configure your Oqtane module to deploy the static assets for Radzen... did you use the debug.cmd in the Deploy project in your external module... or did you create a Nuget package and deploy it?

@sbwalker - I now see that my confusion was, that I thought that somehow, the assets (the .css and .js files) were somehow being _dynamically extracted_ from the .dll assembly. 

I published the normal **Radzen** application project, and I now see that indeed a "**wwwroot/_content/**" folder is created and the assets are placed there:

![image](https://user-images.githubusercontent.com/1857799/103488854-0bfda380-4dc5-11eb-9b68-e5dbbac662e6.png)

Therefore I see your point:

* Unless the assets are placed in this location, Oqtane will not be able to find them.
* With Radzen, I manually extracted the resources from the NuGet package and used code like this, in the debug.cmd file to move the assets: 
`XCOPY "..\Server\wwwroot\Modules\BlazorHelpWebsite.RadzenTest\*" "..\..\oqtaneOne\Oqtane.Server\wwwroot\Modules\BlazorHelpWebsite.RadzenTest\" /Y /S /I`

So I decided put the files here:

![image](https://user-images.githubusercontent.com/1857799/103489552-3b62df00-4dca-11eb-938e-feeb8165f07b.png)

And use this line in the Debug.cmd:

`XCOPY "..\Server\wwwroot\*" "..\..\oqtaneOne\Oqtane.Server\wwwroot\" /Y /S /I`

But, when I try to compile the project I get:

![image](https://user-images.githubusercontent.com/1857799/103489509-dad3a200-4dc9-11eb-805b-1d7e70fa0d87.png)

The project is here: 
[BlazorHelpWebsite.RadzenTest.zip](https://github.com/oqtane/oqtane.framework/files/5762330/BlazorHelpWebsite.RadzenTest.zip)

**Reply by sbwalker** [reply]
I believe the error is caused by the fact that you manually created the wwwroot/_content/Radzen.Blazor folder in your module Server project and when Visual Studio does a build it tries to copy any static assets which exist in linked Nuget packages to the wwwroot/_content folder. So basically Visual Studio is saying that it cannot copy the file from the Radzen.Blazor Nuget package because the file already exists.

**Reply by ADefWebserver** [reply]
> So basically Visual Studio is saying that it cannot copy the file from the Radzen.Blazor Nuget package because the file already exists.

@sbwalker - I thought the same thing. However, if I remove the files I added, yet I have this line in the debug.cmd:

`XCOPY "..\Server\wwwroot\*" "..\..\oqtaneOne\Oqtane.Server\wwwroot\" /Y /S /I`

I assumed that those files would be placed there by the linked NuGet package, and they would be copied to the **Oqtane** solution?
But, no.

As an additional test I renamed my files to add a "2" to the names:

![image](https://user-images.githubusercontent.com/1857799/103494181-52192e00-4dea-11eb-976b-b5ed9c06b680.png)

Yet the **Oqtane** solution only gets my files:

![image](https://user-images.githubusercontent.com/1857799/103494225-97d5f680-4dea-11eb-80af-a9605c286dbd.png)

We have a solution for Radzen, because they will allow us to place the files in the Modules folder.
Syncfusion doesn't allow this :(

**Reply by sbwalker** [reply]
The build error message you posted above "The static web asset ... has a conflicting web root path ..." is because the Radzen package is declared in your Client project and during compilation it notices that if a publish was attempted there would be a conflict because it would try to extract the static resources from the Nuget package to the _content/Radzen.Blazor folder - but the folder already exists in the Server project ( and both folders map to the same physical location during publish ). Its strange that it throws an error rather than a warning but I guess this is what the .NET Core team thought was appropriate.

**Reply by ADefWebserver** [reply]
> is because the Radzen package is declared in your Client project and during compilation it notices that if a publish was attempted there would be a conflict because it would try to extract the static resources from the Nuget package to the _content/Radzen.Blazor folder - but the folder already exists in the Server project ( and both folders map to the same physical location during publish ).

@sbwalker - My initial thinking is "Oh I'll just remove that reference in the Client project..." Then I realize that I actually need that reference (sigh).

**Reply by sbwalker** [reply]
So it looks like this behaviorial breaking change was introduced in .NET 5:

https://github.com/aspnet/Announcements/issues/443

I have found a LOT of references to this online causing grief for developers. Microsoft was trying to resolve a fairly narrow use case - but in the process affected a lot of other use cases which they had not anticipated.

**Reply by sbwalker** [reply]
@ADefWebserver could you try adding the following to your Client or Server project file:

```
  <ItemGroup>
    <Content Remove="wwwroot\_content\**\*.*" />
    <None Include="wwwroot\_content\**\*.*" />
  </ItemGroup>
```

Try to build the solution and see if the "The static web asset ... has a conflicting web root path ..." error still happens.

**Reply by ADefWebserver** [reply]
> @ADefWebserver could you try adding the following to your Client or Server project file:
> 
> ```
>   <ItemGroup>
>     <Content Remove="wwwroot\_content\**\*.*" />
>     <None Include="wwwroot\_content\**\*.*" />
>   </ItemGroup>
> ```
> 
> Try to build the solution and see if the "The static web asset ... has a conflicting web root path ..." error still happens.

@sbwalker - Yeah that appears to work.

This:

![image](https://user-images.githubusercontent.com/1857799/103571535-eb465400-4e7f-11eb-9885-5fcd8bde0fa2.png)

Causes this:

![image](https://user-images.githubusercontent.com/1857799/103571554-f4cfbc00-4e7f-11eb-96a1-26cdd769dc67.png)

But add this (to the **Server** project _only_ of the **Oqtane** module):

![image](https://user-images.githubusercontent.com/1857799/103571634-1761d500-4e80-11eb-84b6-817cf9c46ae1.png)

With a **Debug.cmd** that looks like this:

![image](https://user-images.githubusercontent.com/1857799/103571697-38c2c100-4e80-11eb-81e9-352f90c55871.png)

And it properly builds to the **Oqtane** solution like this:

![image](https://user-images.githubusercontent.com/1857799/103571756-51cb7200-4e80-11eb-8664-d4584e73b1a1.png)

**Reply by ADefWebserver** [reply]
@sbwalker - Ok this weekend I will try to get the **Syncfusion** to work again. Their Nuget package has a *lot* of **.css** and **.js** files, so it would take a while to set up everything...

**Reply by sbwalker** [reply]
This is good news! Essentially it still means that a module developer must manually extract the static assets from the library's Nuget package and put them in the wwwroot/_content folder ( I am guessing the easiest way to do this is to do a Publish on the module Server project to a local folder - which will copy the static assets out of the Nuget packages and you can then copy them into the wwwroot folder ). The *.Server.csproj and debug.cmd template files could be updated in the Module Creator with the changes needed to support this automatically. When building in Release mode, the *.nuspec file appears to be fine already and will create a Nuget package which includes the _content folder... and I believe the installer will also extract the files to the correct locations just fine. So no changes to the Oqtane framework are needed. I like this solution because it should be intuitive for developers and avoids a custom folder naming convention that is specific to Oqtane.

**Reply by ADefWebserver** [reply]
> So no changes to the Oqtane framework are needed. I like this solution because it should be intuitive for developers and avoids a custom folder naming convention that is specific to Oqtane.

@sbwalker  - Yeah if this works (I think  it should) it keeps **Oqtane** out of the way if Microsoft decides to change things down the line.

**Reply by ADefWebserver** [reply]
[BlazorHelpWebsite.SyncfusionTest.zip](https://github.com/oqtane/oqtane.framework/files/5767767/BlazorHelpWebsite.SyncfusionTest.zip)

@sbwalker - I wanted to give you a quick update that this does solve the JavaScript issues with Syncfusion, but it appears there is a problem with the .css

![image](https://user-images.githubusercontent.com/1857799/103599333-c2db4b80-4eb9-11eb-8a3d-b0be1e4cfc50.png)
![image](https://user-images.githubusercontent.com/1857799/103599378-d8507580-4eb9-11eb-8225-052694b97d34.png)
![image](https://user-images.githubusercontent.com/1857799/103599394-e4d4ce00-4eb9-11eb-92ee-395f39fe54d0.png)
![image](https://user-images.githubusercontent.com/1857799/103599415-eef6cc80-4eb9-11eb-9d69-7c997593ef91.png)

I will continue to try to work on this, this weekened...

**Reply by ADefWebserver** [reply]
@sbwalker @larsontim12  - Never mind I had a typo. I got it working 👍 

Working project: 
[BlazorHelpWebsite.SyncfusionTest.zip](https://github.com/oqtane/oqtane.framework/files/5767819/BlazorHelpWebsite.SyncfusionTest.zip)


![image](https://user-images.githubusercontent.com/1857799/103600210-bf48c400-4ebb-11eb-9b87-0363f48ed853.png)
![image](https://user-images.githubusercontent.com/1857799/103600906-60844a00-4ebd-11eb-91f2-71578cdf9ad7.png)

I plan to do a Blog post this weekend.

**Reply by larsontim12** [reply]
@ADefWebserver @sbwalker Awesome!  I will take a look at the sample. Appreciate all of the effort in getting it working!
Thanks,
Tim

**Reply by sbwalker** [reply]
@ADefWebserver Awesome! I made the necessary changes to the external module template in PR #1041

**Reply by ADefWebserver** [reply]
@sbwalker - Great. Will this be released in the next 30 days? If so I would like to wait to do my Blog post. I would want to document the *superior experience* :)

**Reply by ADefWebserver** [reply]
@sbwalker - Actually I see the change is small, so I will go ahead and work on the Blog post this weekend.

**Reply by ADefWebserver** [reply]
I posted a Blog: [Using Syncfusion In Oqtane Modules](https://blazorhelpwebsite.com/ViewBlogPost/47)

**Comment by sbwalker** [comment]
Thank you for all your help in researching this issue - I will take a closer look tomorrow using your example module.

**Reply by ADefWebserver** [reply]
Yeah after re-reading this: [Blazor JavaScript isolation and object references](https://docs.microsoft.com/en-us/aspnet/core/blazor/call-javascript-from-dotnet?view=aspnetcore-5.0#blazor-javascript-isolation-and-object-references) I think a lot of other Component Vendors will start using it.

**Reply by sbwalker** [reply]
I experimented with your example module this morning and I also found some more technical documentation on RCL behavior which explains some of the "magic" which happens behind the scenes when you build or deploy a normal .NET Core project:

https://docs.microsoft.com/en-us/aspnet/core/razor-pages/ui-class?view=aspnetcore-3.0&tabs=visual-studio#consume-content-from-a-referenced-rcl

Like most things in .NET Core, there is only native support for hard-wired ( ie. statically linked ) dependencies. The documentation describes what happens if you have an RCL project as an explicit dependency of your main consuming project - or if you have Nuget package reference to an RCL in your main consuming project. Neither of these scenarios are applicable to Oqtane, as it does not have static dependencies - it relies on dynamic dependencies which are determined at run-time.

**Reply by ADefWebserver** [reply]
@sbwalker - Perhaps a "_brute force_" solution is to:

1. Extract the **.css** and **.js** files from the **Syncfusion** Nuget Package (unlike **Radzen** there area  *lot* of them)
2. Put the files in the **wwwroot/Modules/_content** directory in the **Oqtane** Module Server project
3. In the **Debug.cmd** file, copy the **wwwroot/Modules/_content** files to the **Oqtane wwwroot/_content/{libraryName}/** directory
